test/eqns.lean

```diff
@@ -0,0 +1,38 @@
+import Std.Tactic.Eqns
+import Std.Tactic.GuardMsgs
+
+set_option linter.missingDocs false
+
+def transpose {m n} (A : m → n → Nat) : n → m → Nat
+  | i, j => A j i
+
+theorem transpose_apply {m n} (A : m → n → Nat) (i j) :
+  transpose A i j = A j i := rfl
+
+attribute [eqns transpose_apply] transpose
+
+theorem transpose_const {m n} (c : Nat) :
+    transpose (fun (_i : m) (_j : n) => c) = fun _j _i => c := by
+  fail_if_success rw [transpose]
+  fail_if_success simp [transpose]
+  funext i j
+  rw [transpose]
```

Comment:
At least with lean `v.4.5.0-rc1`, the example works even without the `eqns` attribute above.

Perhaps the heuristics have changed?  Is this extension still useful?

Reply:
I just checked that there are `attribute [eqns]` in Mathlib that have effect / are needed (although some are not needed!).

I don't think there is any particular clamour for this PR: it's just trying to remove tactics from Mathlib that are "needed" before the basic algebraic hierarchy.

I will see if I can move a better test case up from Mathlib, but low priority.

Reply:
@eric-wieser, this is actually the example in the doc-string in Mathlib, but it seems it is no longer relevant there. Would you be willing to rewrite the doc-string?

Reply:
The main use case of `eqns` is actually the `irreducible_def X := ...` command in Mathlib, which makes use of `opaque` to give a completely irreducible definition `X` and a `X_def` theorem, and then uses `attribute [eqns]` to allow `rw [X]` to achieve the same as `rw [X_def]`.